// ios-explore-mcp-server/dist/errors.js
export class IOSExploreStructuredError extends Error {
    source;
    code;
    action;
    baseURL;
    status;
    timeoutMs;
    bodySnippet;
    constructor(error) {
        super(error.message);
        this.name = "IOSExploreStructuredError";
        this.source = error.source;
        if (error.code !== undefined)
            this.code = error.code;
        if (error.action !== undefined)
            this.action = error.action;
        if (error.baseURL !== undefined)
            this.baseURL = error.baseURL;
        if (error.status !== undefined)
            this.status = error.status;
        if (error.timeoutMs !== undefined)
            this.timeoutMs = error.timeoutMs;
        if (error.bodySnippet !== undefined)
            this.bodySnippet = error.bodySnippet;
    }
    toJSON() {
        const result = { source: this.source, message: this.message };
        if (this.code !== undefined)
            result.code = this.code;
        if (this.action !== undefined)
            result.action = this.action;
        if (this.baseURL !== undefined)
            result.baseURL = this.baseURL;
        if (this.status !== undefined)
            result.status = this.status;
        if (this.timeoutMs !== undefined)
            result.timeoutMs = this.timeoutMs;
        if (this.bodySnippet !== undefined)
            result.bodySnippet = this.bodySnippet;
        return result;
    }
}
export function bodySnippet(body) {
    return body.length > 500 ? `${body.slice(0, 500)}...` : body;
}
