// ios-explore-mcp-server/dist/iosExploreClient.js
import { requestTimeoutForAction } from "./config.js";
import { bodySnippet, IOSExploreStructuredError } from "./errors.js";
import { isFailureEnvelope } from "./types.js";
export class IOSExploreClient {
    config;
    constructor(config) {
        this.config = config;
    }
    async call(action, data = {}) {
        const timeoutMs = requestTimeoutForAction(this.config, action, data);
        const controller = new AbortController();
        const timer = setTimeout(() => controller.abort(), timeoutMs);
        let response;
        try {
            response = await fetch(this.config.baseURL, {
                method: "POST",
                headers: { "Content-Type": "application/json" },
                body: JSON.stringify({ action, data }),
                signal: controller.signal
            });
        }
        catch (error) {
            const code = error instanceof Error && error.name === "AbortError" ? "request_timeout" : "connection_failed";
            throw new IOSExploreStructuredError({
                source: "transport",
                code,
                message: error instanceof Error ? error.message : String(error),
                action,
                baseURL: this.config.baseURL,
                timeoutMs
            });
        }
        finally {
            clearTimeout(timer);
        }
        const text = await response.text();
        if (!response.ok) {
            throw new IOSExploreStructuredError({
                source: "http",
                status: response.status,
                message: `HTTP ${response.status}`,
                action,
                bodySnippet: bodySnippet(text)
            });
        }
        let envelope;
        try {
            envelope = JSON.parse(text);
        }
        catch {
            throw new IOSExploreStructuredError({
                source: "http",
                code: "invalid_json",
                message: "HTTP response was not valid JSON",
                action,
                bodySnippet: bodySnippet(text)
            });
        }
        if (isFailureEnvelope(envelope)) {
            throw new IOSExploreStructuredError({
                source: "ios_envelope",
                code: envelope.code,
                message: envelope.message,
                action
            });
        }
        return envelope.data ?? {};
    }
}
