// ios-explore-mcp-server/dist/staticTools.js
import { IOSExploreStructuredError } from "./errors.js";
import { errorResult, jsonResult } from "./result.js";
// ui.inspect 合法可选字段（used for inspectOptions whitelist）。
// Task 3 已从 Swift inputSchema 删除 includeDisabled/includeStaticText/includeContainers，
// 此处同步移除，避免 pickAllowedFields 把 App 不认识的键透传过去。
const inspectOptionKeys = [
    "includeHidden",
    "maxDepth",
    "accessibilityIdentifier",
    "accessibilityIdentifierPrefix",
    "textLimit",
    "maxTargets"
];
// ui.waitAny 合法字段（used for waitAny input whitelist）。
// 等同于 waitAndInspectSchema().properties 的顶层键名，
// 与 App 端 ui.waitAny inputSchema additionalProperties:false 保持一致。
const waitAnyKeys = [
    "conditions",
    "timeoutMs",
    "intervalMs",
    "stableMs",
    "includeHidden"
];
export function createStaticTools(options) {
    const { client, registry } = options;
    return {
        health_check: {
            name: "health_check",
            description: "检查 Mac MCP server 是否能连到 iOSExplore App 的 ping/help。自动加载动态工具。",
            inputSchema: { type: "object", properties: {} },
            handler: async () => {
                try {
                    const ping = await client.call("ping");
                    await client.call("help");
                    // 自动刷新动态工具，避免初次调用时 dynamicToolCount 为 0
                    await registry.refresh();
                    return jsonResult({ ok: true, ping, dynamicToolCount: registry.tools().length, conflicts: registry.conflicts() });
                }
                catch (error) {
                    return jsonResult({ ok: false, error: normalizeError(error), dynamicToolCount: registry.tools().length }, false);
                }
            }
        },
        refresh_tools: {
            name: "refresh_tools",
            description: "重新读取 iOSExplore help 输出并刷新动态 MCP tools。",
            inputSchema: { type: "object", properties: {} },
            handler: async () => {
                const result = await registry.refresh();
                // 与 health_check 对齐：统一用 dynamicToolCount 表达"动态工具数量"，
                // 不再混用 toolCount/dynamicToolCount 两套字段名（之前 health_check 用
                // dynamicToolCount 而 refresh_tools 返回 toolCount，调用方难以一致消费）。
                return jsonResult({
                    dynamicToolCount: result.toolCount,
                    conflicts: result.conflicts,
                    ...(result.error ? { error: result.error } : {})
                });
            }
        },
        call_action: {
            name: "call_action",
            description: "兜底转发任意 iOSExplore action。优先使用固定工具或动态工具；排障时使用本工具。",
            inputSchema: {
                type: "object",
                properties: {
                    action: { type: "string" },
                    data: { type: "object" }
                },
                required: ["action"]
            },
            handler: async (input) => {
                const action = String(input.action ?? "");
                if (!action) {
                    return errorResult({ source: "mcp_server", code: "missing_action", message: "call_action requires an 'action' field" });
                }
                const data = objectValue(input.data);
                try {
                    return jsonResult(await client.call(action, data));
                }
                catch (error) {
                    // transport 失败做一次重试，仍失败就附上 healthCheck + nextSteps，
                    // 与动态工具（server.ts:callTool）的 transport 错误模式对齐，
                    // 让 Agent 在兜底工具上也能拿到相同的排障上下文（不再是裸的 connection_failed）。
                    if (isTransportError(error)) {
                        await sleep(200);
                        try {
                            return jsonResult(await client.call(action, data));
                        }
                        catch (retryError) {
                            if (isTransportError(retryError)) {
                                return errorResult(await enrichTransportError(retryError, client));
                            }
                            return resultForFailure(normalizeError(retryError));
                        }
                    }
                    return resultForFailure(normalizeError(error));
                }
            }
        },
        wait_and_inspect: {
            name: "wait_and_inspect",
            description: "先调用 ui.waitAny，再调用 ui.inspect。wait_timeout 后仍尽量返回最新 observation。",
            inputSchema: waitAndInspectSchema(),
            handler: async (input) => {
                const inspectOptions = pickAllowedFields(objectValue(input.inspectOptions), inspectOptionKeys);
                try {
                    const wait = await client.call("ui.waitAny", pickAllowedFields(input, waitAnyKeys));
                    const observation = await client.call("ui.inspect", inspectOptions);
                    return jsonResult({ wait, observation });
                }
                catch (error) {
                    const normalized = normalizeError(error);
                    if (normalized.source === "ios_envelope" && normalized.code === "wait_timeout") {
                        const observation = await client.call("ui.inspect", inspectOptions);
                        return jsonResult({ wait: normalized, observation }, false);
                    }
                    return errorResult(normalized);
                }
            }
        },
        ui_wait: {
            name: "ui_wait",
            description: "等待 UI 稳定或等待目标/文本/快照变化",
            inputSchema: uiWaitSchema(),
            handler: async (input) => {
                try {
                    return jsonResult(await client.call("ui.wait", input));
                }
                catch (error) {
                    return resultForFailure(normalizeError(error));
                }
            }
        },
        ui_tap_and_inspect: {
            name: "ui_tap_and_inspect",
            description: "点击元素后自动检查 UI 状态。组合 ui.tap + ui.wait + ui.inspect，减少 Agent 推理次数。95% 的点击场景需要检查结果，建议优先使用此工具。",
            inputSchema: uiTapAndInspectSchema(),
            handler: async (input) => {
                const startTime = Date.now();
                const tapParams = pickAllowedFields(input, ["accessibilityIdentifier", "path", "viewSnapshotID"]);
                const waitForStable = input.waitForStable !== false; // default true
                const stableTimeMs = typeof input.stableTimeMs === "number" ? input.stableTimeMs : 300;
                const inspectDepth = typeof input.inspectDepth === "number" ? input.inspectDepth : 2;
                const inspectMaxTargets = typeof input.inspectMaxTargets === "number" ? input.inspectMaxTargets : 20;
                const timing = {
                    tapMs: 0,
                    inspectMs: 0,
                    totalMs: 0
                };
                try {
                    // Step 1: Execute tap
                    const tapStart = Date.now();
                    const tapResult = await client.call("ui.tap", tapParams);
                    timing.tapMs = Date.now() - tapStart;
                    // Step 2: Wait for UI to stabilize (if requested)
                    if (waitForStable) {
                        const waitStart = Date.now();
                        try {
                            await client.call("ui.wait", { mode: "idle", stableMs: stableTimeMs, timeoutMs: stableTimeMs + 1000 });
                            timing.waitMs = Date.now() - waitStart;
                        }
                        catch (waitError) {
                            // If wait times out, continue to inspect anyway
                            timing.waitMs = Date.now() - waitStart;
                        }
                    }
                    // Step 3: Inspect current UI state
                    const inspectStart = Date.now();
                    const inspectResult = await client.call("ui.inspect", {
                        maxDepth: inspectDepth,
                        maxTargets: inspectMaxTargets
                    });
                    timing.inspectMs = Date.now() - inspectStart;
                    timing.totalMs = Date.now() - startTime;
                    return jsonResult({
                        tap: tapResult,
                        stateAfter: inspectResult,
                        timing
                    });
                }
                catch (error) {
                    // If tap fails, return error immediately without continuing
                    const normalized = normalizeError(error);
                    timing.totalMs = Date.now() - startTime;
                    return errorResult({ ...normalized, timing });
                }
            }
        }
    };
}
function waitAndInspectSchema() {
    return {
        type: "object",
        properties: {
            conditions: {
                type: "array",
                minItems: 1,
                maxItems: 16,
                description: "每项必须包含 id/mode；targetExists/targetGone 需要 accessibilityIdentifier 或 path；textExists 需要 text；snapshotChanged 需要 viewSnapshotID；idle 无额外字段。"
            },
            timeoutMs: { type: "number" },
            intervalMs: { type: "number" },
            stableMs: { type: "number" },
            includeHidden: { type: "boolean" },
            inspectOptions: {
                type: "object",
                description: "传给 ui.inspect 的可选参数。只能传 ui.inspect 真实字段（includeHidden / maxDepth / accessibilityIdentifier / accessibilityIdentifierPrefix / textLimit / maxTargets），不接受 ui.topViewHierarchy 专用的 detailLevel 或 ui.waitAny 专用的 conditions 等字段。",
                properties: {
                    includeHidden: { type: "boolean" },
                    maxDepth: { type: "integer" },
                    accessibilityIdentifier: { type: "string" },
                    accessibilityIdentifierPrefix: { type: "string" },
                    textLimit: { type: "integer" },
                    maxTargets: { type: "integer" }
                },
                additionalProperties: false
            }
        },
        required: ["conditions"]
    };
}
function uiWaitSchema() {
    return {
        type: "object",
        properties: {
            mode: {
                type: "string",
                enum: ["idle", "targetExists", "targetGone", "textExists", "snapshotChanged"],
                description: "等待模式: idle / targetExists / targetGone / textExists / snapshotChanged"
            },
            timeoutMs: {
                type: "integer",
                minimum: 0,
                maximum: 30000,
                description: "业务超时毫秒数, 范围 0...30000, 默认 3000"
            },
            intervalMs: {
                type: "integer",
                minimum: 50,
                maximum: 5000,
                description: "轮询间隔毫秒数, 范围 50...5000, 默认 100"
            },
            stableMs: {
                type: "integer",
                minimum: 0,
                maximum: 10000,
                description: "idle 模式下连续稳定的毫秒数, 范围 0...10000, 默认 300"
            },
            text: {
                type: "string",
                description: "textExists 模式要等待的文本片段"
            },
            viewSnapshotID: {
                type: "string",
                description: "snapshotChanged 模式参照的 viewSnapshotID (由 ui.inspect 签发)"
            },
            accessibilityIdentifier: {
                type: "string",
                description: "targetExists/targetGone 模式: 按 accessibilityIdentifier 精确定位目标 view"
            },
            path: {
                type: "string",
                description: "targetExists/targetGone 模式: 按 ui.inspect 或 ui.topViewHierarchy 返回的 root/0/1 路径定位目标 view"
            },
            includeHidden: {
                type: "boolean",
                description: "idle/textExists/targetExists/targetGone 是否考虑隐藏 view, 默认 false"
            }
        }
    };
}
function uiTapAndInspectSchema() {
    return {
        type: "object",
        properties: {
            accessibilityIdentifier: {
                type: "string",
                description: "按 accessibilityIdentifier 精确定位目标 view。与 path 二选一（互斥）：两字段中有且仅提供一个"
            },
            path: {
                type: "string",
                description: "按 ui.inspect 或 ui.topViewHierarchy 返回的 root/0/1 路径定位目标 view。与 accessibilityIdentifier 二选一（互斥）：两字段中有且仅提供一个"
            },
            viewSnapshotID: {
                type: "string",
                description: "ui.inspect 签发的结构化 target 指纹快照标识"
            },
            waitForStable: {
                type: "boolean",
                description: "是否等待 UI 稳定后再 inspect，默认 true"
            },
            stableTimeMs: {
                type: "integer",
                minimum: 0,
                maximum: 3000,
                description: "等待 UI 稳定的时长（毫秒），默认 300"
            },
            inspectDepth: {
                type: "integer",
                minimum: 0,
                maximum: 10,
                description: "inspect 的最大递归深度，默认 2"
            },
            inspectMaxTargets: {
                type: "integer",
                minimum: 1,
                maximum: 512,
                description: "inspect 返回的最大目标数量，默认 20"
            }
        },
        required: ["viewSnapshotID"]
    };
}
function objectValue(value) {
    return typeof value === "object" && value !== null && !Array.isArray(value) ? value : {};
}
function withoutKey(input, key) {
    const copy = { ...input };
    delete copy[key];
    return copy;
}
function pickAllowedFields(input, keys) {
    const allowed = new Set(keys);
    return Object.fromEntries(Object.entries(input).filter(([key]) => allowed.has(key)));
}
function resultForFailure(error) {
    if (error.source === "ios_envelope") {
        // call_action 的 unknown_action 保持 isError=false（兜底工具本就让 agent 猜 action 名）；
        // invalid_data / stale_locator 升格为 isError=true（参数错误、snapshot 过期）；
        // 其它 ios_envelope code（wait_timeout / alert_unavailable 等）保持 isError=false。
        const errorCodes = ["invalid_data", "stale_locator"];
        if (error.code && errorCodes.includes(error.code)) {
            return errorResult(error);
        }
        return jsonResult(error, false);
    }
    return errorResult(error);
}
function normalizeError(error) {
    if (error instanceof IOSExploreStructuredError) {
        return error.toJSON();
    }
    // 有明确 source 的对象，保留原始 source（transport/http/ios_envelope），
    // 不全部覆盖为 ios_envelope。只在没有 source 时默认 ios_envelope。
    if (typeof error === "object" && error !== null && "source" in error && "code" in error) {
        const err = error;
        if (err.source === "transport" || err.source === "http" || err.source === "ios_envelope") {
            return { source: err.source, code: err.code, message: err.message ?? String(error) };
        }
        return { source: "ios_envelope", code: err.code, message: err.message ?? String(error) };
    }
    return {
        source: "mcp_server",
        code: "unexpected_error",
        message: error instanceof Error ? error.message : String(error)
    };
}
// transport 错误判定与 server.ts 中动态工具路径一致，避免兜底工具走另一套判定。
function isTransportError(error) {
    return error instanceof IOSExploreStructuredError && error.source === "transport";
}
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}
// 与 server.ts 中 enrichTransportError 对齐：兜底工具的 transport 失败也带
// retry/healthCheck/nextSteps 三段排障上下文，调用方拿到的错误结构在不同入口上一致。
async function enrichTransportError(error, caller) {
    const normalized = error.toJSON();
    return {
        ...normalized,
        retry: { attempted: true, delayMs: 200, succeeded: false },
        healthCheck: await pingHealthCheck(caller),
        nextSteps: [
            "iOSExplore App 当前不可达；如果是真机调试，请确认 App 仍在运行、iproxy 仍在监听 38321，并用 XcodeBuildMCP launch_app_device 以 IOS_EXPLORE_AUTOSTART=1 重启后再试。"
        ]
    };
}
async function pingHealthCheck(caller) {
    try {
        return { ok: true, ping: await caller.call("ping") };
    }
    catch (error) {
        return { ok: false, error: normalizeError(error) };
    }
}
