// ios-explore-mcp-server/dist/toolName.js
export function toolNameForAction(action) {
    return action.replace(/[^A-Za-z0-9_]/g, "_");
}
export function buildActionToolMap(commands, fixedToolNames) {
    const grouped = new Map();
    for (const command of commands) {
        const name = toolNameForAction(command.action);
        const existing = grouped.get(name) ?? [];
        existing.push(command);
        grouped.set(name, existing);
    }
    const tools = [];
    const conflicts = [];
    for (const [toolName, entries] of grouped) {
        if (entries.length > 1 || fixedToolNames.has(toolName)) {
            conflicts.push({ toolName, actions: entries.map(entry => entry.action) });
            continue;
        }
        const entry = entries[0];
        if (!entry) {
            continue;
        }
        tools.push({
            name: toolName,
            description: `${entry.description}\n\nOriginal iOSExplore action: ${entry.action}`,
            inputSchema: entry.inputSchema,
            action: entry.action
        });
    }
    return { tools, conflicts };
}
